document-store/src/main/java/org/hypertrace/core/documentstore/mongo/MongoCollection.java

```diff
@@ -130,6 +131,30 @@ public boolean upsert(Key key, Document document) throws IOException {
     }
   }
 
+  /** Same as existing upsert method, however, extends the support with condition */
+  @Override
+  public boolean upsert(Key key, Document document, Filter condition, @Nullable Boolean isUpsert)
```

Comment:
Since we are adding a new api should we also have a better response object encapsulating success/failure, number of docs upserted, any further error desc in case of errors etc etc

Reply:
Done